src/Features/Core/Diagnostics/EngineV1/DiagnosticIncrementalAnalyzer.StateManager.cs

```diff
@@ -0,0 +1,236 @@
+                return _hostStates.GetOrCreateStateSets(project.Language).Concat(_projectStates.GetOrUpdateStateSets(project));
+            }
+
+            /// <summary>
+            /// Return <see cref="StateSet"/>s for the given <see cref="Project"/>. 
+            /// This will either return already created <see cref="StateSet"/>s for the specific snapshot of <see cref="Project"/> or
+            /// It will create new <see cref="StateSet"/>s for the <see cref="Project"/>.
+            /// Unlike <see cref="GetOrUpdateStateSets(Project)"/>, this has no side effect.
+            /// </summary>
+            public IEnumerable<StateSet> GetOrCreateStateSets(Project project)
+            {
+                return _hostStates.GetOrCreateStateSets(project.Language).Concat(_projectStates.GetOrCreateStateSets(project));
+            }
+
+            /// <summary>
+            /// Return <see cref="StateSet"/> for the given <see cref="DiagnosticAnalyzer"/> in the context of <see cref="Project"/>.
+            /// This will either return already created <see cref="StateSet"/> for the specific snapshot of <see cref="Project"/> or
+            /// It will create new <see cref="StateSet"/> for the <see cref="Project"/>.
+            /// This will not have any side effect.
+            /// </summary>
+            public StateSet GetOrCreateStateSet(Project project, DiagnosticAnalyzer analyzer)
+            {
+                var stateSet = _hostStates.GetOrCreateStateSet(project.Language, analyzer);
+                if (stateSet != null)
+                {
+                    return stateSet;
+                }
+
+                return _projectStates.GetOrCreateStateSet(project, analyzer);
+            }
+
+            /// <summary>
+            /// Remove given <see cref="ProjectId"/>.
+            /// </summary>
+            public void RemoveStateSet(ProjectId projectId)
+            {
+                _projectStates.RemoveStateSet(projectId);
+            }
+
+            private void RaiseProjectAnalyzerReferenceChanged(ProjectAnalyzerReferenceChangedEventArgs args)
+            {
+                ProjectAnalyzerReferenceChanged?.Invoke(this, args);
+            }
+
+            private static ImmutableDictionary<DiagnosticAnalyzer, StateSet> CreateAnalyzerMap(
+                    string language, IEnumerable<ImmutableArray<DiagnosticAnalyzer>> analyzerCollection)
+            {
+                var builder = ImmutableDictionary.CreateBuilder<DiagnosticAnalyzer, StateSet>();
+                foreach (var analyzers in analyzerCollection)
+                {
+                    foreach (var analyzer in analyzers)
+                    {
+                        // TODO: 
+                        // #1, all de -duplication should move to HostAnalyzerManager
+                        // #2, not sure whether de-duplicatoin of analyzer itself makes sense. this can only happen
+                        //     if user delibrately put same analyzer twice.
```

Comment:
ya, the test you have put some comment above about de-duplication. I think if a user deliberately did that, they will get duplicated diagnostic.